scalableminds/webknossos
# Redesign change user experience modal

In this PR the modal to change the experience of one or multiple users got a rework.
Now there is a difference between editing multiple users' and a single user's experience.

### single user
- The modal used for editing a single user can be opened by clicking on one of the experience tag in the user table. Alternatively it can be opened by selecting only one user and pressing the "Change Experience"-button.
 - The modal should consist of two components. The first component is a table that lists all current experience domains of the user with their value. The first column shows the name of the experience domain. The second one shows the value in an editable NumberInput-field which accepts only values greater than 0. Editing the value of a experience domain makes a icon right next to the input-field visible. Pressing this Icon should reset the value to the initial value. This feature is only available to domains that already existed before starting editing. With the column on the right it is possible to remove a experience domain and also to undo this change. Removed entries should be disabled (except for the undo-icon).
- If the table has more than 5 entries it, gets a scrollbar. With the scrollbar enabled antd changes the whole layout of the table (but this should be acceptable / I don't know how to fix this).
- The second component is a input-field in which you can enter multiple new domains. Next to it is a add-button to add all entered domains. If a domain is entered, that is already listed it will be deleted from the input at the moment it is entered, so that there will never be two domains with the same name in the table. Also entering a domain with less than 3 characters is considered invalid => the input gets rejected and a toast should appear to inform the user.
- The modal should not be closeable by clicking on the background. Pressing the submit button will submit all changes and the cancel button just closes the modal. Each time the modal gets visible the changes previous entered will be erased.

### muliple users
- Has three components:
- A table that display all shared experience domains. If the value of each user of a domain is equal, only the value will be displayed in the value-column. Otherwise the text "varying from [lowest_value] to [highest_value]" will be displayed.
- The next component is almost identical to the editing table in the single user's modal. All changes will be listed here. Adding a domain can be done via the third component the input-field. The difference here is that this table is smaller (height) than in the single user version. All users selected will be changed according to these entries. A listed domain, that is not marked as deleted will result in all users having this domain with the specified value when submitting. All users that have a domain that is marked as removed will have this domain removed after submitting. Also on the very right side of the table there is a button to remove the entered domain completely from the changes-table to make it possible to remove accidental entered changes.
- The select-component for domains is the same as in the single users version but also has a "remove domains"-button that adds the entered domains to the change-table but directly set as removed.
-Submitting the entered changes will result in the changes described in the third point.

### general:
After submitting the changes the user-list-view should not fetch the users' data again. Instead it should use the result of the backend send by changing the experiences of a user.


~~~=> Feature to discuss and to implement: Adding a domain in the select-input with the following pattern: "(domain_name -> text): (value -> number)" will be directly parsed into a table entry with the entered domain_name and its value.~~~ 

# Edit 
## changes of the rework
- there is only one table in the modal. The shared experience table (showed when multiple users are sleected) got removed. Instead a new column was added to the other table to indicate the range of shared experience domains.
- while editing the exp of many users you can check a checkbox to only show and edit the shared  experiences of all users.
- clicking the trash icon will remove the entry completely
- there is no more "remove experience"-button 
- just entering a experience in the select field will add it to the table
- revert changes button still exist
- the title of the table got removed

### URL of deployed dev instance (used for testing):
- [https://redesignchangeuserexperiencemodal.webknossos.xyz](https://redesignchangeuserexperiencemodal.webknossos.xyz)

### Steps to test:
- open the users tab of the admins views
- change experience domains as you like from a single user and a multiple users
- test if the usage is somehow unintuitive
- check for unintended behavior 

### Issues:
- fixes [#3033](https://github.com/scalableminds/webknossos/issues/3033)

------
- ~~~[ ] Updated [migration guide](../blob/master/MIGRATIONS.md#unreleased) if applicable~~~
- ~~~[ ] Needs datastore update after deployment~~~
